### PR TITLE
Gatekeeper views

### PR DESCRIPTION
**Pull request by Jakub89**
[#7](https://github.com/golemfactory/concent/issues/7)

### CODE REVIEW

File: concent_api/concent_api/settings/development.py [hunk 1]
```diff
@@ -18,3 +18,5 @@
     "gatekeeper",
     "admin-panel",
 ]
+
+STORAGE_CLUSTER_ADDRESS = 'http://storage.concent.golem.network/'
```
Comment (cameel) [inline concent_api/concent_api/settings/development.py]: Set the default to `'http://devel.concent.golem.network/'`. storage.concent.golem.network Does not exist yet.

File: concent_api/concent_api/system_check.py [hunk 2]
```diff
@@ -40,3 +42,17 @@ def check_settings_concent_features(app_configs, **kwargs):  # pylint: disable=u
             ))
 
     return errors_and_warnings
+
+
+@register()
+def check_settings_storage_cluster_address(app_configs, **kwargs):  # pylint: disable=unused-argument
+    storage_cluster_address = URLValidator(schemes=['http', 'https'])
```
Comment (cameel) [inline concent_api/concent_api/system_check.py]: This variable does not store cluster address so why is it called `storage_cluster_address`? It's totally misleading. This should be called something like `url_validator`.

Please always call things for what they really are. It makes the code easier to read.

File: concent_api/requirements.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 Django >= 1.11, < 1.12
 psycopg2
--e git://github.com/golemfactory/golem-messages.git@v1.1.0#egg=golem_messages
+-e git://github.com/Code-Poets/golem-messages.git@303ff545e9db7631b078ada2f8a9a4b336c39b30#egg=Golem_Messages
```
Comment (cameel) [inline concent_api/requirements.txt]: It's preferable to use branch name in `requirements.txt`. Unless you really really need only this particular commit which, I think, is not the case here. `requirements.lock` will get a commit anyway.

File: concent_api/gatekeeper/views.py [hunk 4]
```diff
@@ -26,9 +30,11 @@ def upload(request):
     init_messages()
     response = parse_headers(request)
     if response is not None:
+        logger.warning(response.content.decode())
         return response
+    logger.info('Request passed all upload validations.')
 
-    return HttpResponse("", status = 200)
+    return HttpResponse("Request passed all upload validations.", status = 200)
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: All responses should be in the JSON format.

File: concent_api/gatekeeper/views.py [hunk 5]
```diff
@@ -40,9 +46,11 @@ def download(request):
     init_messages()
     response = parse_headers(request)
     if response is not None:
+        logger.warning(response.content.decode())
         return response
+    logger.info('Request passed all download validations.')
 
-    return HttpResponse("", status = 200)
+    return HttpResponse("Request passed all download validations.", status = 200)
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: All responses should be in the JSON format.

File: concent_api/gatekeeper/views.py [hunk 6]
```diff
@@ -74,4 +78,59 @@ def parse_headers(request):
...
+    if not all(isinstance(path, str) for path in loaded_golem_message.files):
+        return gatekeeper_access_denied_response('Wrong type of files variable.')
+    if request.method == 'POST':
+        path_to_file = request.get_full_path().partition("/upload-auth/")[2]
+    if request.method == 'GET':
+        path_to_file = request.get_full_path().partition("/download-auth/")[2]
+    if path_to_file not in loaded_golem_message.files:
+        return gatekeeper_access_denied_response('Path to specified file is not listed in files variable.')
+
+    # TMP check if header doesn't contains Concent-Upload-Path or X-Session-ID:
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: All `TMP`s should be resolved before sumitting a pull request.

File: concent_api/gatekeeper/views.py [hunk 7]
```diff
@@ -74,4 +78,59 @@ def parse_headers(request):
...
+    if request.method == 'GET' and loaded_golem_message.operation != 'download':
+        return gatekeeper_access_denied_response('Wrong operation variable for this request method.')
+
+    # -FILES
+    if not all(isinstance(path, str) for path in loaded_golem_message.files):
+        return gatekeeper_access_denied_response('Wrong type of files variable.')
+    if request.method == 'POST':
+        path_to_file = request.get_full_path().partition("/upload-auth/")[2]
+    if request.method == 'GET':
+        path_to_file = request.get_full_path().partition("/download-auth/")[2]
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: Don't hard-code URLs. Use `reverse()`.

File: concent_api/gatekeeper/views.py [hunk 8]
```diff
@@ -74,4 +78,59 @@ def parse_headers(request):
...
+        return gatekeeper_access_denied_response('Wrong type of token_expiration_deadline variable.')
+    current_time = int(datetime.datetime.now().timestamp())
+
+    if current_time > loaded_golem_message.token_expiration_deadline:
+        return gatekeeper_access_denied_response('token_expiration_deadline has passed.')
+
+    # -STORAGE_CLUSTER_ADDRESS
+    if not isinstance(loaded_golem_message.storage_cluster_address, str):
+        return gatekeeper_access_denied_response('Wrong type of storage_cluster_address variable.')
+    cluster_address = URLValidator()
```
Comment (cameel) [inline concent_api/gatekeeper/views.py]: This should be called `url_validator`. It's not an address.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        self.assertEqual(response3.status_code, 401)
+        self.assertIn('message', response3.json().keys())
+
+    @freeze_time("2017-11-17 10:00:00")
+    def test_download_should_return_401_if_wrong_token_cluster_address(self):
+        upload_token = self.upload_token
+        upload_token.storage_cluster_address = 'www://storage.concent.golem.network/'
+        golem_upload_token = dump(upload_token, settings.CONCENT_PRIVATE_KEY, settings.CONCENT_PUBLIC_KEY)
+        encrypted_token = b64encode(golem_upload_token).decode()
+        response = self.client.get('/gatekeeper/download-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend',
```
Comment (cameel) [inline concent_api/gatekeeper/tests/test_gatekeeper_views.py]: Don't hard-code paths. The `/download-auth/` part should be taken from a call to `reverse()`.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.assertIsInstance(response3, JsonResponse)
+        self.assertEqual(response3.status_code, 401)
+        self.assertIn('message', response3.json().keys())
+
+    @freeze_time("2017-11-17 10:00:00")
+    def test_download_should_return_401_if_wrong_token_cluster_address(self):
+        upload_token = self.upload_token
+        upload_token.storage_cluster_address = 'www://storage.concent.golem.network/'
+        golem_upload_token = dump(upload_token, settings.CONCENT_PRIVATE_KEY, settings.CONCENT_PUBLIC_KEY)
```
Comment (cameel) [inline concent_api/gatekeeper/tests/test_gatekeeper_views.py]: You should not rely on setting values in tests. If you're using any settings, hard-code their values with `override_settings()`.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        golem_upload_token = dump(self.upload_token, settings.CONCENT_PRIVATE_KEY, settings.CONCENT_PUBLIC_KEY)
+        encrypted_token = b64encode(golem_upload_token).decode()
+        response1 = self.client.get('/gatekeeper/download-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend',
+            HTTP_AUTHORIZATION = 'GolemGolem ' + encrypted_token,
+            HTTP_CONCENT_CLIENT_PUBLIC_KEY = self.public_key
+        )
+
+        self.assertIsInstance(response1, JsonResponse)
+        self.assertEqual(response1.status_code, 401)
+        self.assertIn('message', response1.json().keys())
```
Comment (cameel) [inline concent_api/gatekeeper/tests/test_gatekeeper_views.py]: You could just do it three times in a loop and put all the values of `HTTP_AUTHORIZATION` your're testing on a list. It would also make it easier for you to test more cases. For example you should also test: empty header, header consisting of just whitespace, header with more than one space, etc.

**Comment by cameel** [comment]
There's one small change in the spec: https://github.com/golemfactory/concent/issues/7#issuecomment-355044525

It's pretty straightforward so please implement it as a part of this pull request.

File: concent_api/concent_api/settings/development.py (hunk 1 above)
Comment (Jakub89) [reply concent_api/concent_api/settings/development.py]: Done

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py (hunk 11 above)
Comment (Jakub89) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: I've made a loop with `wrong_test_headers = ['GolemGolem '+ encrypted_token, '', 'Golem encrypted_token ', 'Golem '+ 'a' + encrypted_token]`, but I left response2 to check if HTTP_AUTHORIZATION header exists at all.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py (hunk 10 above)
Comment (Jakub89) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: Done. Added `@override_settings(
    CONCENT_PRIVATE_KEY = b'l\xcdh\x19\xeb$>\xbcG\xa1\xc7v\xe8\xd7o\x0c\xbf\x0e\x0fM\x89lw\x1e\xd7K\xd6Hnv$\xa2',
    CONCENT_PUBLIC_KEY  = b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;'
)` before each class.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: Done. Switch to:
```
response = Client().post(
            '{}{}'.format(
                reverse('gatekeeper:upload'),
                'blender/benchmark/test_task/scene-Helicopter-27-cycles.blend'
            ),
```

File: concent_api/gatekeeper/views.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/gatekeeper/views.py]: Okey, now it's `return JsonResponse({"message": "Request passed all upload validations."}, status = 200)`

File: concent_api/gatekeeper/views.py (hunk 5 above)
Comment (Jakub89) [reply concent_api/gatekeeper/views.py]: Okey, now it's `return JsonResponse({"message": "Request passed all download validations."}, status = 200)`

File: concent_api/gatekeeper/views.py (hunk 8 above)
Comment (Jakub89) [reply concent_api/gatekeeper/views.py]: Done

File: concent_api/gatekeeper/views.py (hunk 7 above)
Comment (Jakub89) [reply concent_api/gatekeeper/views.py]: Switched to:
```
    if request.method == 'POST':
        path_to_file = request.get_full_path().partition(reverse('gatekeeper:upload'))[2]
    if request.method == 'GET':
        path_to_file = request.get_full_path().partition(reverse('gatekeeper:download'))[2]
```

File: concent_api/gatekeeper/views.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/gatekeeper/views.py]: I've deleted this part.

**Comment by Jakub89** [comment]
I've rebase both branches `gatekeeper-app` and then `gatekeeper-views` on @kbeker 's `valid-golem-messages`. Everything seems to work fine.

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py (hunk 11 above)
Comment (cameel) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: Oh, so you added additional `N` at the end of the header name to test with an invalid header. You should use someting more obvious. Like `HTTP_AUTHORIZATION_ABC`. I did not even notice that.

You can make the loop cover this too by defining a list of dicts instead of a list of strings:

``` python
wrong_test_headers = [
    {'HTTP_AUTHORIZATION':  'GolemGolem ' + encrypted_token},
    {'HTTP_AUTHORIZATIONN': 'GolemGolem ' + encrypted_token},
    {'HTTP_AUTHORIZATION':  ''},
    {'HTTP_AUTHORIZATION':  'Golem encrypted_token '},
    {'HTTP_AUTHORIZATION':  'Golem a' + encrypted_token},
]
```

Then pass the headers just like `kwargs`:
``` python
for headers in wrong_test_headers:
    response = Client().get(..., **headers)
    ...
```

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py (hunk 10 above)
Comment (cameel) [reply concent_api/gatekeeper/tests/test_gatekeeper_views.py]: What about `STORAGE_CLUSTER_ADDRESS`? All the settings you rely on should be overridden. Not only those you refer to explicitly.

File: concent_api/requirements.txt (hunk 3 above)
Comment (cameel) [reply concent_api/requirements.txt]: Now you have no version at all. You should add a branch name.

Comment (Jakub89) [reply concent_api/requirements.txt]: FileTransferToken has been pulled to master, so I changed version to:
`-e git://github.com/golemfactory/golem-messages.git@master#egg=golem_messages`

File: concent_api/gatekeeper/tests/test_gatekeeper_views.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        self.upload_token.files[0]['path']      = 'blender/benchmark/test_task/scene-Helicopter-27-cycles.blend'
+        self.upload_token.files[0]['checksum']  = '098f6bcd4621d373cade4e832627b4f6'
+        self.upload_token.files[0]['size']      = 1024
+        self.upload_token.operation             = 'upload'
+
+    @freeze_time("2018-12-30 11:00:00")
+    def test_upload_should_accept_valid_message(self):
+        golem_upload_token = dump(self.upload_token, settings.CONCENT_PRIVATE_KEY, settings.CONCENT_PUBLIC_KEY)
+        encrypted_token = b64encode(golem_upload_token).decode()
+        response = Client().post(
```
Comment (cameel) [inline concent_api/gatekeeper/tests/test_gatekeeper_views.py]: You don't have to create a new `Client()` instance yourself. It already exists. Django creates `self.client` for you automatically.
